### DISCUSSION TITLE
[Fastlane.Swift] Integrating Third Party SPM Dependencies With Restorable Lanes

### DISCUSSION THREAD

**Discussion by TimOliver** [fastlane/fastlane#18511] [Ideas] [answered]
Hey folks (and @minuscorp)! This is a follow-up idea I'm having to [my previous discussion](https://github.com/fastlane/fastlane/discussions/18483). 

In summary, with the goal of keeping my git history as clean as possible, I helped submit a PR to allow Fastlane Swift to re-install all of its dependent Swift and Xcode project files, so only files like `fastlane/Fastfile.swift` are actually needed to be checked in.

Following on from this, something that would add a *lot* of power to executing Fastlane lanes in Swift is if we could add SPM modules on top of this setup. As an example, I'd like to import [Ink](https://github.com/JohnSundell/Ink) as a dependency so I can convert the markdown in a changelog file into HTML in one of my lanes.

One way to solve this would be to manually edit `FastlaneSwiftRunner.xcodeproj` to add the SPM dependencies normally that way, but then that would require checking in the `xcodeproj` and potentially put it out of sync with the other associated Swift files. And maybe this is the best kind of solution we can have for this sort of setup.

Ideally, I'm wondering if there's a way to specify a list of desired SPM modules we'd want to use with our Fastlane Swift setup, and have them inserted into the `FastlaneSwiftRunner.xcodeproj` whenever it's created/regenerated before running a lane, that would be fantastic. I'm not sure if I'm on the right track here, but if it was as simple as defining a `fastlane/Package.swift` definition, that would be incredible.

Please let me know what you think, and how feasible this is. Thanks! :D

**Comment by minuscorp** [comment, marked as answer]
I think this is already supported out of the box, as no xcode project is needed for this:

1. Create an executable Swift Package Manager project with `swift package init --type executable`.

2. Add an entry point or a main.swift file.
```swift
import Fastlane

// Create a class with: 
class FastFile: LaneFile {}
```

3. Don't forget to start the _fastlane_ runloop as follows:
```swift
import Fastlane

Main().run(with: Fastfile())
```

4. Add lanes to the class.

5. Modify the target of your executable to have executable arguments `lane myLane` or add them in the call after making `swift build`.

6. You can edit the created Package.swift file to add your desired dependencies so you can use them in the Fastfile.

If you want to just push your Package.swift and Package.resolved to the repo, you'd need to `swift build` the package to create your executable again which can be found in the `.debug` or `.release` folders, depending on how you built the package (`.debug` by default).


EDIT: I actually created this PR as I forgot to update the docs once I integrated with SPM a few months ago... https://github.com/fastlane/docs/pull/1044

**Reply by TimOliver** [reply]
Woah! Thanks so much for the extremely detailed description @minuscorp!

Ahhh, okay. I see! So it _is_ possible, but it basically requires making a small Swift project from scratch, and importing Fastlane as another SPM dependency as opposed to just using the regular Fastlane and Swift.

Hmmm. I tried following these steps, but I definitely got a bit stuck in places. 

1. I tried generating an Xcode project with `swift package generate-xcodeproj` but I would consistently get an error: cyclic dependency declaration found: TargetName -> TargetName doing that.
2. I tried to follow the instructions more closely, but when I did a `swift build`, I got the error `could not build Objective-C module 'Fastlane'`

It's possible I'm missing some extra steps in the tutorial that are being assumed. If there's anything I can do to help there, please let me know. :)

**Reply by minuscorp** [reply]
I have this little example in my machine that might help a bit:

```swift
// swift-tools-version:5.2
// The swift-tools-version declares the minimum version of Swift required to build this package.

import PackageDescription

let package = Package(
    name: "fastlaneRunner",
    products: [
        .executable(name: "fastlaneRunner", targets: ["fastlaneRunner"])
    ],
    dependencies: [
        // Dependencies declare other packages that this package depends on.
        .package(name: "Fastlane", url: "https://github.com/fastlane/fastlane", from: "2.179.0")
    ],
    targets: [
        // Targets are the basic building blocks of a package. A target can define a module or a test suite.
        // Targets can depend on other targets in this package, and on products in packages which this package depends on.
        .target(
            name: "fastlaneRunner",
            dependencies: ["Fastlane"],
            path: "Sources/Thingy"
        )
    ]
)
```

I have expanded step 2 on the PR to include this little example.

**Reply by TimOliver** [reply]
Ahhh! I wasn't adding "Fastlane" to the `dependencies` property in the `target`. That explains it. Yep, that example helps A LOT. Thanks so much for that! :D

**Reply by minuscorp** [reply]
The PR with the instructions has been merged so it now can be found on the docs 😄
